src/js/utils/template-parts.js

```diff
@@ -35,24 +36,35 @@ export const defaultProcessor = {
   },
 };
 
+/**
+ * @extends {DocumentFragment}
+ */
 export class TemplateInstance extends window.DocumentFragment {
   #parts;
   #processor;
-  constructor(template, state, processor = defaultProcessor) {
+
+  constructor(
+    /** @type HTMLTemplateElement */ template,
+    /** @type Record<string, any> */ state,
+    processor = defaultProcessor
+  ) {
     super();
+
     this.append(template.content.cloneNode(true));
     this.#parts = parse(this);
+
     this.#processor = processor;
     processor.createCallback?.(this, this.#parts, state);
     processor.processCallback(this, this.#parts, state);
   }
-  update(state) {
+
+  update(/** @type Record<string, any> */ state) {
     this.#processor.processCallback(this, this.#parts, state);
   }
 }
 
 // collect element parts
-export const parse = (element, parts = []) => {
+export const parse = (element, /** @type [string, Part][] */ parts = []) => {
```

Comment:
suggestion(non-blocking): consider adding a `@typedef` for your `PartsTuple` and/or `PartsTupleArray` (See: https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#typedef-callback-and-param)